### PR TITLE
'typeof' BuiltInTypeName|UserDefinedTypeName

### PR DESCRIPTION
**Pull request by djkevincr**
This PR introduces 'typeof' operator to BuiltInTypeName | UserDefinedTypeName.

Eg:-
type intType = typeof int;
type stringType = typeof string;

This enables following type comparisons.

int variable;
if ( (typeof variable) == (typeof int) ) {
// do logic
} else {
// do logic
}

### CODE REVIEW

File: modules/ballerina-lang/src/main/java/org/ballerinalang/model/tree/NodeKind.java [hunk 1]
```diff
@@ -59,6 +59,7 @@
     SIMPLE_VARIABLE_REF,
     STRING_TEMPLATE_LITERAL,
     TERNARY_EXPR,
+    TYPE_ACCESS_EXPR,
```
Comment (sameera-jayasoma) [inline modules/ballerina-lang/src/main/java/org/ballerinalang/model/tree/NodeKind.java]: TYPEOF_EXPRESSION

File: modules/ballerina-lang/src/main/java/org/ballerinalang/model/tree/expressions/TypeAccessExpressionNode.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+package org.ballerinalang.model.tree.expressions;
+
+import org.ballerinalang.model.tree.types.TypeNode;
+
+/**
+ * typeof string
+ *
+ * @since 0.95
+ */
+public interface TypeAccessExpressionNode extends ExpressionNode {
```
Comment (sameera-jayasoma) [inline modules/ballerina-lang/src/main/java/org/ballerinalang/model/tree/expressions/TypeAccessExpressionNode.java]: Let's rename this class to TypeofExpressionNode

File: modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 3]
```diff
@@ -823,6 +828,11 @@ public void visit(BLangRetry retryNode) {
         result = retryNode;
     }
 
+    @Override
+    public void visit(BLangTypeAccessExpr accessExpr) {
```
Comment (sameera-jayasoma) [inline modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: Let's rename this to BLangTypeofExpression

File: modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 4]
```diff
@@ -502,18 +504,40 @@ public void visit(BLangBinaryExpr binaryExpr) {
...
     }
 
+    public void visit(BLangTypeAccessExpr accessExpr) {
+        BType actualType = symTable.typeType;
+        accessExpr.resolvedType = symResolver.resolveTypeNode(accessExpr.typeNode, env);
+        resultTypes = types.checkTypes(accessExpr, Lists.of(actualType), expTypes);
+    }
+
     public void visit(BLangUnaryExpr unaryExpr) {
-        BType exprType = checkExpr(unaryExpr.expr, env).get(0);
```
Comment (sameera-jayasoma) [inline modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java]: I think we can refactor this method. We check for (OperatorKind.TYPEOF.equals(unaryExpr.operator) twice in this method.

File: modules/ballerina-lang/src/main/java/org/ballerinalang/model/tree/NodeKind.java (hunk 1 above)
Comment (djkevincr) [reply modules/ballerina-lang/src/main/java/org/ballerinalang/model/tree/NodeKind.java]: +1

File: modules/ballerina-lang/src/main/java/org/ballerinalang/model/tree/expressions/TypeAccessExpressionNode.java (hunk 2 above)
Comment (djkevincr) [reply modules/ballerina-lang/src/main/java/org/ballerinalang/model/tree/expressions/TypeAccessExpressionNode.java]: +1

File: modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java (hunk 3 above)
Comment (djkevincr) [reply modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: +1

File: modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java (hunk 4 above)
Comment (djkevincr) [reply modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java]: +1

**Pull request merged**